src/libAtomVM/nifs.c

```diff
@@ -3272,6 +3279,97 @@ static term nif_binary_split(Context *ctx, int argc, term argv[])
     return result_list;
 }
 
+static term nif_binary_replace(Context *ctx, int argc, term argv[])
+{
+    term bin_term = argv[0];
+    term pattern = argv[1];
+    term replacement = argv[2];
+    term options = argc == 4 ? argv[3] : term_nil();
+
+    VALIDATE_VALUE(bin_term, term_is_binary);
+    VALIDATE_VALUE(pattern, term_is_binary);
+    VALIDATE_VALUE(replacement, term_is_binary);
+    VALIDATE_VALUE(options, term_is_list);
+
+    bool global = false;
+    while (term_is_nonempty_list(options)) {
+        term head = term_get_list_head(options);
+        if (LIKELY(head == GLOBAL_ATOM)) {
+            global = true;
+        } else {
+            RAISE_ERROR(BADARG_ATOM);
+        }
+        options = term_get_list_tail(options);
+    }
+
+    size_t bin_size = term_binary_size(bin_term);
+    size_t pattern_size = term_binary_size(pattern);
+    size_t repl_size = term_binary_size(replacement);
+
+    if (bin_size < pattern_size || bin_size == 0) {
+        return bin_term;
+    }
+
+    if (UNLIKELY(pattern_size == 0)) {
+        RAISE_ERROR(BADARG_ATOM);
+    }
+
+    const char *bin_data = term_binary_data(bin_term);
+    const char *pattern_data = term_binary_data(pattern);
+    const char *repl_data = term_binary_data(replacement);
+
```

Comment:
This entire block can be replaced with memmem on systems that have it.
Just add a TODO for it.